File: WinForms/VB/TreeListMainDemo/CodeExamples/ColumnsAndBands.vb
Imports System.Windows.Forms
Imports DevExpress.DXperience.Demos.CodeDemo
Imports DevExpress.DXperience.Demos.CodeDemo.Data
Imports DevExpress.XtraEditors
Imports DevExpress.XtraTreeList
Imports DevExpress.XtraTreeList.Columns

Namespace Examples

    <DevExpress.DXperience.Demos.CodeDemo.CodeExampleClass("Columns and bands", "ColumnsAndBands.cs")>
    Public Module ColumnsAndBands

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleSetUp>
        Public Function SetUp(ByVal sampleHost As DevExpress.XtraEditors.XtraUserControl) As Object()
            Dim treeList As DevExpress.XtraTreeList.TreeList = New DevExpress.XtraTreeList.TreeList()
            treeList.Dock = System.Windows.Forms.DockStyle.Fill
            treeList.Parent = sampleHost
            AddHandler treeList.Load, Sub(s, e) treeList.ExpandAll()
            Return New Object() {treeList}
        End Function

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleTearDown>
        Public Sub TearDown(ByVal sampleHost As DevExpress.XtraEditors.XtraUserControl)
            For Each control As System.Windows.Forms.Control In sampleHost.Controls
                control.Dispose()
            Next

            sampleHost.Controls.Clear()
        End Sub

        '
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Add columns manually"), Examples.SpaceObjectCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("AddVisible")>
        Public Sub AddColumnsManually(ByVal treeList As DevExpress.XtraTreeList.TreeList)
            Dim colName As DevExpress.XtraTreeList.Columns.TreeListColumn = treeList.Columns.AddVisible(fieldName:="Name")
            ' Prohibit edit the column's cells
            colName.OptionsColumn.AllowEdit = False
            ' Use this property to override the TreeList.OptionsCustomization.AllowColumnResizing property's value 
            colName.OptionsColumn.AllowSize = False
            Dim colTypeOfObject As DevExpress.XtraTreeList.Columns.TreeListColumn = treeList.Columns.AddVisible(fieldName:="TypeOfObject")
            ' Use this property to override the TreeList.OptionsCustomization.AllowColumnMoving property's value
            colTypeOfObject.OptionsColumn.AllowMove = False
            ' Specifies whether columns can be moved by dragging their headers.
            treeList.OptionsCustomization.AllowColumnMoving = True
            ' Specifies whether columns can be resized by dragging the right edge of their headers.
            treeList.OptionsCustomization.AllowColumnResizing = True
            ' There are no new columns will be added here because the columns set is already predefined
            treeList.DataSource = DevExpress.DXperience.Demos.CodeDemo.Data.SpaceObject.GetData()
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Add bands manually"), Examples.SpaceObjectCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("AddBand", "Add")>
        Public Sub Bands(ByVal treeList As DevExpress.XtraTreeList.TreeList)
            ' All the columns will be created automatically
            treeList.DataSource = DevExpress.DXperience.Demos.CodeDemo.Data.SpaceObject.GetData()
            ' Adding bands
            Dim mainBand As DevExpress.XtraTreeList.Columns.TreeListBand = treeList.Bands.AddBand(caption:="Main")
            Dim detailsBand As DevExpress.XtraTreeList.Columns.TreeListBand = treeList.Bands.AddBand(caption:="Details")
            ' Adding sub-bands
            Dim markBand As DevExpress.XtraTreeList.Columns.TreeListBand = detailsBand.Bands.AddBand(caption:="Is Checked")
            Dim addInfoBand As DevExpress.XtraTreeList.Columns.TreeListBand = detailsBand.Bands.AddBand(caption:="Additional info")
            ' Adding columns into bands
            mainBand.Columns.Add(treeList.Columns("Name"))
            mainBand.Columns.Add(treeList.Columns("TypeOfObject"))
            markBand.Columns.Add(treeList.Columns("Mark"))
            addInfoBand.Columns.Add(treeList.Columns("Notes"))
            addInfoBand.Columns.Add(treeList.Columns("RecordDate"))
            addInfoBand.Columns.Add(treeList.Columns("MeanRadiusByEarth"))
            addInfoBand.Columns.Add(treeList.Columns("MeanRadiusInKM"))
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Fixed columns"), Examples.SpaceObjectCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("Fixed", "FixedLineWidth")>
        Public Sub FixedColumns(ByVal treeList As DevExpress.XtraTreeList.TreeList)
            ' All the columns will be created automatically
            treeList.DataSource = DevExpress.DXperience.Demos.CodeDemo.Data.SpaceObject.GetData()
            treeList.OptionsView.AutoWidth = False
            treeList.Columns(CStr(("Notes"))).Width = 384
            ' Make the specific columns fixed
            treeList.Columns(CStr(("Name"))).Fixed = DevExpress.XtraTreeList.Columns.FixedStyle.Left
            treeList.Columns(CStr(("TypeOfObject"))).Fixed = DevExpress.XtraTreeList.Columns.FixedStyle.Right
            ' Setup the fixed line width
            treeList.FixedLineWidth = 2
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Fixed bands"), Examples.SpaceObjectCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("Fixed", "FixedLineWidth")>
        Public Sub FixedBands(ByVal treeList As DevExpress.XtraTreeList.TreeList)
            ' All the columns will be created automatically
            treeList.DataSource = DevExpress.DXperience.Demos.CodeDemo.Data.SpaceObject.GetData()
            treeList.OptionsView.AutoWidth = False
            treeList.Columns(CStr(("Notes"))).Width = 700
            ' Creating bands
            Dim mainBand As DevExpress.XtraTreeList.Columns.TreeListBand = treeList.Bands.AddBand("Main")
            mainBand.Columns.Add(treeList.Columns("Name"))
            Dim detailsBand As DevExpress.XtraTreeList.Columns.TreeListBand = treeList.Bands.AddBand("Details")
            detailsBand.Columns.Add(treeList.Columns("TypeOfObject"))
            detailsBand.Columns.Add(treeList.Columns("Notes"))
            Dim InfoBand As DevExpress.XtraTreeList.Columns.TreeListBand = treeList.Bands.AddBand("Information")
            InfoBand.Columns.Add(treeList.Columns("Mark"))
            InfoBand.Columns.Add(treeList.Columns("RecordDate"))
            ' Make the specific band fixed
            mainBand.Fixed = DevExpress.XtraTreeList.Columns.FixedStyle.Left
            InfoBand.Fixed = DevExpress.XtraTreeList.Columns.FixedStyle.Right
            ' Setup the fixed line width
            treeList.FixedLineWidth = 2
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Column separators"), Examples.SpaceObjectCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("SeparatorWidth")>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleVersionID(181)>
        Public Sub SeparatorColumns(ByVal treeList As DevExpress.XtraTreeList.TreeList)
            ' All the columns will be created automatically
            treeList.DataSource = DevExpress.DXperience.Demos.CodeDemo.Data.SpaceObject.GetData()
            treeList.OptionsView.AutoWidth = True
            ' Enable a separator for a column
            treeList.Columns(CStr(("MeanRadiusInKM"))).SeparatorWidth = 2
            treeList.Appearance.Separator.BackColor = System.Drawing.Color.LightPink
        End Sub

        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleCase("Band separators"), Examples.SpaceObjectCS>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleUnderlineTokens("SeparatorWidth")>
        <DevExpress.DXperience.Demos.CodeDemo.CodeExampleVersionID(181)>
        Public Sub SeparatorBands(ByVal treeList As DevExpress.XtraTreeList.TreeList)
            ' All the columns will be created automatically
            treeList.DataSource = DevExpress.DXperience.Demos.CodeDemo.Data.SpaceObject.GetData()
            treeList.OptionsView.AutoWidth = True
            ' Creating bands
            Dim mainBand As DevExpress.XtraTreeList.Columns.TreeListBand = treeList.Bands.AddBand("Main")
            mainBand.Columns.Add(treeList.Columns("Name"))
            Dim detailsBand As DevExpress.XtraTreeList.Columns.TreeListBand = treeList.Bands.AddBand("Details")
            detailsBand.Columns.Add(treeList.Columns("TypeOfObject"))
            detailsBand.Columns.Add(treeList.Columns("MeanRadiusInKM"))
            detailsBand.Columns.Add(treeList.Columns("Notes"))
            Dim InfoBand As DevExpress.XtraTreeList.Columns.TreeListBand = treeList.Bands.AddBand("Information")
            InfoBand.Columns.Add(treeList.Columns("Mark"))
            detailsBand.Columns.Add(treeList.Columns("MeanRadiusByEarth"))
            InfoBand.Columns.Add(treeList.Columns("RecordDate"))
            ' Enable separators for a column and band
            detailsBand.SeparatorWidth = 2
            treeList.Columns(CStr(("MeanRadiusInKM"))).SeparatorWidth = 1
            treeList.Appearance.Separator.BackColor = System.Drawing.Color.LightPink
        End Sub
    End Module
End Namespace

File: WinForms/CS/TreeListMainDemo/CodeExamples/ColumnsAndBands.cs
namespace Examples {
    using System.Windows.Forms;
    using DevExpress.DXperience.Demos.CodeDemo;
    using DevExpress.DXperience.Demos.CodeDemo.Data;
    using DevExpress.XtraEditors;
    using DevExpress.XtraTreeList;
    using DevExpress.XtraTreeList.Columns;

    [CodeExampleClass("Columns and bands", "ColumnsAndBands.cs")]
    public static class ColumnsAndBands {
        [CodeExampleSetUp]
        public static object[] SetUp(XtraUserControl sampleHost) {
            TreeList treeList = new TreeList();
            treeList.Dock = DockStyle.Fill;
            treeList.Parent = sampleHost;
            treeList.Load += (s, e) =>
            {
                treeList.ExpandAll();
            };
            return new object[] { treeList };
        }
        [CodeExampleTearDown]
        public static void TearDown(XtraUserControl sampleHost) {
            foreach(Control control in sampleHost.Controls)
                control.Dispose();
            sampleHost.Controls.Clear();
        }
        //
        [CodeExampleCase("Add columns manually"), SpaceObjectCS]
        [CodeExampleUnderlineTokens("AddVisible")]
        public static void AddColumnsManually(TreeList treeList) {
            TreeListColumn colName = treeList.Columns.AddVisible(fieldName: "Name");
            // Prohibit edit the column's cells
            colName.OptionsColumn.AllowEdit = false;
            // Use this property to override the TreeList.OptionsCustomization.AllowColumnResizing property's value 
            colName.OptionsColumn.AllowSize = false;

            TreeListColumn colTypeOfObject = treeList.Columns.AddVisible(fieldName: "TypeOfObject");
            // Use this property to override the TreeList.OptionsCustomization.AllowColumnMoving property's value
            colTypeOfObject.OptionsColumn.AllowMove = false;

            // Specifies whether columns can be moved by dragging their headers.
            treeList.OptionsCustomization.AllowColumnMoving = true;
            // Specifies whether columns can be resized by dragging the right edge of their headers.
            treeList.OptionsCustomization.AllowColumnResizing = true;
            // There are no new columns will be added here because the columns set is already predefined
            treeList.DataSource = SpaceObject.GetData();
        }
        [CodeExampleCase("Add bands manually"), SpaceObjectCS]
        [CodeExampleUnderlineTokens("AddBand", "Add")]
        public static void Bands(TreeList treeList) {
            // All the columns will be created automatically
            treeList.DataSource = SpaceObject.GetData();
            // Adding bands
            TreeListBand mainBand = treeList.Bands.AddBand(caption: "Main");
            TreeListBand detailsBand = treeList.Bands.AddBand(caption: "Details");
            // Adding sub-bands
            TreeListBand markBand = detailsBand.Bands.AddBand(caption: "Is Checked");
            TreeListBand addInfoBand = detailsBand.Bands.AddBand(caption: "Additional info");
            // Adding columns into bands
            mainBand.Columns.Add(treeList.Columns["Name"]);
            mainBand.Columns.Add(treeList.Columns["TypeOfObject"]);
            markBand.Columns.Add(treeList.Columns["Mark"]);
            addInfoBand.Columns.Add(treeList.Columns["Notes"]);
            addInfoBand.Columns.Add(treeList.Columns["RecordDate"]);
            addInfoBand.Columns.Add(treeList.Columns["MeanRadiusByEarth"]);
            addInfoBand.Columns.Add(treeList.Columns["MeanRadiusInKM"]);
        }
        [CodeExampleCase("Fixed columns"), SpaceObjectCS]
        [CodeExampleUnderlineTokens("Fixed", "FixedLineWidth")]
        public static void FixedColumns(TreeList treeList) {
            // All the columns will be created automatically
            treeList.DataSource = SpaceObject.GetData();
            treeList.OptionsView.AutoWidth = false;
            treeList.Columns["Notes"].Width = 384;
            // Make the specific columns fixed
            treeList.Columns["Name"].Fixed = FixedStyle.Left;
            treeList.Columns["TypeOfObject"].Fixed = FixedStyle.Right;
            // Setup the fixed line width
            treeList.FixedLineWidth = 2;
        }
        [CodeExampleCase("Fixed bands"), SpaceObjectCS]
        [CodeExampleUnderlineTokens("Fixed", "FixedLineWidth")]
        public static void FixedBands(TreeList treeList) {
            // All the columns will be created automatically
            treeList.DataSource = SpaceObject.GetData();
            treeList.OptionsView.AutoWidth = false;
            treeList.Columns["Notes"].Width = 700;
            // Creating bands
            TreeListBand mainBand = treeList.Bands.AddBand("Main");
            mainBand.Columns.Add(treeList.Columns["Name"]);
            TreeListBand detailsBand = treeList.Bands.AddBand("Details");
            detailsBand.Columns.Add(treeList.Columns["TypeOfObject"]);
            detailsBand.Columns.Add(treeList.Columns["Notes"]);            
            TreeListBand InfoBand = treeList.Bands.AddBand("Information");
            InfoBand.Columns.Add(treeList.Columns["Mark"]);
            InfoBand.Columns.Add(treeList.Columns["RecordDate"]);
            // Make the specific band fixed
            mainBand.Fixed = FixedStyle.Left;
            InfoBand.Fixed = FixedStyle.Right;
            // Setup the fixed line width
            treeList.FixedLineWidth = 2;
        }
        [CodeExampleCase("Column separators"), SpaceObjectCS]
        [CodeExampleUnderlineTokens("SeparatorWidth")]
        [CodeExampleVersionID(181)]
        public static void SeparatorColumns(TreeList treeList) {
            // All the columns will be created automatically
            treeList.DataSource = SpaceObject.GetData();
            treeList.OptionsView.AutoWidth = true;
            // Enable a separator for a column
            treeList.Columns["MeanRadiusInKM"].SeparatorWidth = 2;
            treeList.Appearance.Separator.BackColor = System.Drawing.Color.LightPink;
        }
        [CodeExampleCase("Band separators"), SpaceObjectCS]
        [CodeExampleUnderlineTokens("SeparatorWidth")]
        [CodeExampleVersionID(181)]
        public static void SeparatorBands(TreeList treeList) {
            // All the columns will be created automatically
            treeList.DataSource = SpaceObject.GetData();
            treeList.OptionsView.AutoWidth = true;
            // Creating bands
            TreeListBand mainBand = treeList.Bands.AddBand("Main");
            mainBand.Columns.Add(treeList.Columns["Name"]);
            TreeListBand detailsBand = treeList.Bands.AddBand("Details");
            detailsBand.Columns.Add(treeList.Columns["TypeOfObject"]);
            detailsBand.Columns.Add(treeList.Columns["MeanRadiusInKM"]);
            detailsBand.Columns.Add(treeList.Columns["Notes"]);
            TreeListBand InfoBand = treeList.Bands.AddBand("Information");
            InfoBand.Columns.Add(treeList.Columns["Mark"]);
            detailsBand.Columns.Add(treeList.Columns["MeanRadiusByEarth"]);
            InfoBand.Columns.Add(treeList.Columns["RecordDate"]);
            // Enable separators for a column and band
            detailsBand.SeparatorWidth = 2;
            treeList.Columns["MeanRadiusInKM"].SeparatorWidth = 1;
            treeList.Appearance.Separator.BackColor = System.Drawing.Color.LightPink;
        }
    }
}
